### PR TITLE
MAT-4510 - Add "Add Stratification" Button

### PR DESCRIPTION
**Pull request by ethankaplan**
## MADiE PR

Jira Ticket: [MAT-4510](https://jira.cms.gov/browse/MAT-4510)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [x] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).
* [ ] All CDN/Web dependencies are hosted internally (i.e MADiE-Root Repo).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/components/measureGroups/MeasureGroups.tsx [hunk 1]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                          value={formik.values.stratifications[i].cqlDefinition}
+                          inputProps={{
+                            "data-testid": "stratification-select",
+                          }}
+                          onChange={(e) => {
+                            formik.setFieldValue(
+                              `stratifications[${i}]cqlDefinition`,
+                              e.target.value
+                            );
+                          }}
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx]: You can avoid manual binding.
`onChange={formik.handleChange}` would be sufficient

File: src/components/measureGroups/MeasureGroups.tsx [hunk 2]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                          InputLabelProps={{ shrink: false }}
+                          SelectProps={{
+                            native: true,
+                          }}
+                          name="type"
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx]: I think this should be 
```name={`stratifications[${i}].cqlDefinition`}```

File: src/components/measureGroups/MeasureGroups.tsx [hunk 3]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                          value={formik.values.stratifications[i].association}
+                          inputProps={{
+                            "data-testid": "association-select",
+                          }}
+                          onChange={(e) => {
+                            formik.setFieldValue(
+                              `stratifications[${i}].association`,
+                              e.target.value
+                            );
+                          }}
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx]: `onChange={formik.handleChange}`

File: src/components/measureGroups/MeasureGroups.tsx [hunk 4]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                          InputLabelProps={{ shrink: false }}
+                          SelectProps={{
+                            native: true,
+                          }}
+                          name="type"
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx]: ```name={`stratifications[${i}].association`}```

File: src/components/measureGroups/MeasureGroups.tsx [hunk 5]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                          Stratification {i + 1} Description
+                        </FieldLabel>
+                        <FieldSeparator>
+                          {canEdit && (
+                            <textarea
+                              value={
+                                formik.values.stratifications[i].description
+                              }
+                              //type="text"
+                              name="stratification-description"
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx]: ```name={`stratifications[${i}].description`}```

File: src/components/measureGroups/MeasureGroups.tsx [hunk 6]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                              </option>
+                            ))}
+                        </TextField>
+                      </Col>
+                      <Col>
+                        <FieldLabel htmlFor="stratification-description">
+                          Stratification {i + 1} Description
+                        </FieldLabel>
+                        <FieldSeparator>
+                          {canEdit && (
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx:705]: this field will not be visible if you don't own the measure. I think we would want it to be read only in that case.

File: src/components/measureGroups/MeasureGroups.tsx [hunk 7]
```diff
@@ -640,114 +640,141 @@ const MeasureGroups = () => {
...
+                    data-testid="add-strat-button"
+                    onClick={(e) => {
+                      e.preventDefault();
+                      formik.values.stratifications = [
+                        ...formik.values.stratifications,
+                        EmptyStrat,
+                      ];
+
+                      //idk how to force a component update in a less dumb way
+                      if (successMessage !== "") {
```
Comment (adongare) [inline src/components/measureGroups/MeasureGroups.tsx:740]: Will leave up to you but I feel like this is a good usecase of Formik FieldArray. it gives you a bunch of helpers that you can use to add/remove form fields(avoid manual handling) into array of records, in this case stratification array:
https://codesandbox.io/s/formik-fieldarray-forked-lyygcg?file=/src/index.js:1660-1668

https://formik.org/docs/api/fieldarray

File: src/components/measureGroups/MeasureGroups.tsx (hunk 6 above)
Comment (ethankaplan) [reply src/components/measureGroups/MeasureGroups.tsx:705]: That's not really in scope of this ticket. Should it be corrected here? Or should we open up another ticket for this?

File: src/components/measureGroups/MeasureGroups.tsx (hunk 7 above)
Comment (ethankaplan) [reply src/components/measureGroups/MeasureGroups.tsx:740]: I intentionally designed it for 4511 with removing stratifications. I'll do research on this and if it's better I'll change it when both are integrated.

File: src/components/measureGroups/MeasureGroups.tsx (hunk 6 above)
Comment (adongare) [reply src/components/measureGroups/MeasureGroups.tsx:705]: We can create a new ticket

**Review by adongare — APPROVED** [review]

**Pull request merged**